lib/libzfs/libzfs_pool.c

```diff
@@ -4536,6 +4587,24 @@ zpool_obj_to_path(zpool_handle_t *zhp, uint64_t dsobj, uint64_t obj,
 	free(mntpnt);
 }
 
+/*
+ * Given an dataset object number, return data block and indirect block size.
+ */
+int
+zpool_get_block_size(zpool_handle_t *zhp, uint64_t obj,
```

Comment:
Ok, this would explain why my test case was working in some cases but not others.  That is, it worked when corrupted files in `/testpool` but not in `/testpool/testdir`

Reply:
I am working on changes! I shall update pull request today. We can test it afterwards.